### PR TITLE
[ssh] enable user's saved ssh public keys

### PR DESCRIPTION
**Pull request by suiguoxin**
https://github.com/microsoft/pai/issues/5274

Test cases:

- generate a pair of  ssh keys, save the public keys to rest-server
``` json 
PUT rest-api/api/v2/users/me
body: 
{
    "data": {
        "username": "admin",
        "extension": {
            "sshKeys": [
                 { 
                      "title": "test-title", 
                      "value": "your-ssh-public-key" 
                 }
            ]
        }
    },
    "patch": true
}
```
- create a test job, enable user ssh, add a new public key
- test ssh connection with both the pre saved key and the new public key

### CODE REVIEW

File: src/plugins/ssh/init.py [hunk 1]
```diff
@@ -49,9 +70,20 @@ def main():
...
     if "userssh" in parameters:
-        if "type" in parameters["userssh"] and "value" in parameters["userssh"]:
+        # get user public keys from rest server
+        application_token = plugin_config.get("application_token")
+        username = os.environ.get("PAI_USER_NAME")
+        public_keys = []
+        if application_token:
+            public_keys = get_user_public_keys(application_token, username)
+
+        if "value" in parameters["userssh"]:
```
Comment (hzy46) [inline src/plugins/ssh/init.py:80]: ` if "value" in parameters["userssh"] and parameters["userssh"]["value"] != ""`

Comment (hzy46) [reply src/plugins/ssh/init.py:80]: After we modified webportal, `parameters["userssh"]["value"]` could be `""`

**Review by hzy46 — APPROVED** [review]

File: src/plugins/ssh/init.py (hunk 1 above)
Comment (suiguoxin) [reply src/plugins/ssh/init.py]: > After we modified webportal, `parameters["userssh"]["value"]` could be `""`

Fixed, thx

File: src/plugins/ssh/init.py [hunk 2]
```diff
@@ -49,9 +86,20 @@ def main():
     cmd_params = [jobssh]
 
     if "userssh" in parameters:
-        if "type" in parameters["userssh"] and "value" in parameters["userssh"]:
+        # get user public keys from rest server
+        application_token = plugin_config.get("application_token")
+        username = os.environ.get("PAI_USER_NAME")
+        public_keys = []
+        if application_token:
```
Comment (Binyang2014) [inline src/plugins/ssh/init.py:93]: Please catch except and return an error code and log something here

**Review by Binyang2014 — APPROVED** [review]

File: src/plugins/ssh/init.py (hunk 2 above)
Comment (suiguoxin) [reply src/plugins/ssh/init.py:95]: > Please catch except and return an error code and log something here

Added

File: src/init [hunk 3]
```diff
@@ -63,6 +64,10 @@ function exit_handler()
     exit 254
   fi
 
+  if [[ $CHILD_PROCESS = "PLUGIN_INITIALIZER" ]]; then
+    exit 255
+  fi
+
```
Comment (Binyang2014) [inline src/init:70]: Please remove these lines. I will add error handling in another PR.

File: src/plugins/ssh/init.py [hunk 4]
```diff
@@ -49,9 +88,24 @@ def main():
...
+            try:
+                public_keys = get_user_public_keys(application_token, username)
+            except Exception:  #pylint: disable=broad-except
+                LOGGER.error("Failed to get user public keys", exc_info=True)
+                raise UnknownError("Failed to get user public keys")
```
[the remark this answers is not in the document]
Comment (suiguoxin) [reply src/plugins/ssh/init.py]: > Please just return a error code here. Such as `os.exit(1)`

Fixed

File: src/init (hunk 3 above)
Comment (suiguoxin) [reply src/init]: > Please remove these lines. I will add error handling in another PR.

Removed

**Review by Binyang2014 — APPROVED** [review]

**Pull request merged**